# <center>Разведочный анализ данных (EDA) и предобработка данных</center>

In [357]:
# Импорт библиотеки pandas для работы с табличными данными.
import pandas as pd

In [358]:
# Открываем данные.
project_df = pd.read_csv('data/undefind.csv', sep=';')
project_df

ID истории болезни  \
0      2e1d0b3f-488a-11ed-ab5a-0050568844e6   
1      2e1d0b3f-488a-11ed-ab5a-0050568844e6   
2      2e1d0b3f-488a-11ed-ab5a-0050568844e6   
3      2e1d0b3f-488a-11ed-ab5a-0050568844e6   
4      2e1d0b3f-488a-11ed-ab5a-0050568844e6   
...                                     ...   
36292  516f652e-8bb8-11eb-bb95-2cea7fe73f75   
36293  516f652e-8bb8-11eb-bb95-2cea7fe73f75   
36294  516f652e-8bb8-11eb-bb95-2cea7fe73f75   
36295  516f652e-8bb8-11eb-bb95-2cea7fe73f75   
36296  516f652e-8bb8-11eb-bb95-2cea7fe73f75   

              Осн. диаг. при выписке МКБ10 (текст) Заголовок документа  \
0                           Плазмоклеточный лейкоз  Общий анализ крови   
1                           Плазмоклеточный лейкоз  Общий анализ крови   
2                           Плазмоклеточный лейкоз  Общий анализ крови   
3                           Плазмоклеточный лейкоз  Общий анализ крови   
4                           Плазмоклеточный лейкоз  Общий анализ крови   
...                                            ...                 ...   
36292  Острый лейкоз неуточненного клеточного типа  Общий анализ крови   
36293  Острый лейкоз неуточненного клеточного типа  Общий анализ крови   
36294  Острый лейкоз неуточненного клеточного типа  Общий анализ крови   
36295  Острый лейкоз неуточненного клеточного типа  Общий анализ крови   
36296  Острый лейкоз неуточненного клеточного типа  Общий анализ крови   

                 Кол. лаб. показатель  Значение кол. показателя  \
0                     Лейкоциты (WBC)                      1.80   
1                    Эритроциты (RBC)                      4.18   
2                    Гемоглобин (HGB)                    120.00   
3                    Гематокрит (HCT)                     35.10   
4      Средний объем эритроцита (MCV)                     84.00   
...                               ...                       ...   
36292                        Базофилы                      1.00   
36293                      Эозинофилы                      0.00   
36294                    Пролимфоциты                      0.00   
36295                       Лимфоциты                     65.00   
36296                        Моноциты                      1.00   

      Ед. изм. кол. показателя Норма кол. показателя  \
0                       10*9/л                  4:11   
1                      10*12/л               3,8:5,7   
2                          г/л               117:160   
3                            %                 35:47   
4                           фл                 80:99   
...                        ...                   ...   
36292                        %                   0:2   
36293                        %                   0:5   
36294                        %                   0:0   
36295                        %                 18:44   
36296                        %                  2:12   

      Флаг нормы кол. показателя Кач. лаб. показатель  \
0                          Пониж                  NaN   
1                           Норм                  NaN   
2                           Норм                  NaN   
3                           Норм                  NaN   
4                           Норм                  NaN   
...                          ...                  ...   
36292                       Норм          Комментарий   
36293                       Норм          Комментарий   
36294                       Норм          Комментарий   
36295                      Повыш          Комментарий   
36296                      Пониж          Комментарий   

                                Значение кач. показателя  \
0                                                    NaN   
1                                                    NaN   
2                                                    NaN   
3                                                    NaN   
4                                                    NaN   
...                                             

In [359]:
# Проверяем на дубликаты.
project_df.duplicated().sum()

4091

In [360]:
# Избавимся от дубликатов.
project_df = project_df.drop_duplicates()
project_df.shape

(32206, 13)

In [361]:
# Посмотрим информацию о признаках.
project_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 32206 entries, 0 to 36296
Data columns (total 13 columns):
 #   Column                                Non-Null Count  Dtype  
---  ------                                --------------  -----  
 0   ID истории болезни                    32206 non-null  object 
 1   Осн. диаг. при выписке МКБ10 (текст)  32206 non-null  object 
 2   Заголовок документа                   32206 non-null  object 
 3   Кол. лаб. показатель                  32155 non-null  object 
 4   Значение кол. показателя              32155 non-null  float64
 5   Ед. изм. кол. показателя              32155 non-null  object 
 6   Норма кол. показателя                 32155 non-null  object 
 7   Флаг нормы кол. показателя            32155 non-null  object 
 8   Кач. лаб. показатель                  6931 non-null   object 
 9   Значение кач. показателя              6931 non-null   object 
 10  Норма кач. показателя                 6931 non-null   object 
 11  Пол                 

In [362]:
# Выведем кол-во пропущенных значений в каждом признаке.
project_df.isna().sum()

ID истории болезни                          0
Осн. диаг. при выписке МКБ10 (текст)        0
Заголовок документа                         0
Кол. лаб. показатель                       51
Значение кол. показателя                   51
Ед. изм. кол. показателя                   51
Норма кол. показателя                      51
Флаг нормы кол. показателя                 51
Кач. лаб. показатель                    25275
Значение кач. показателя                25275
Норма кач. показателя                   25275
Пол                                         0
Дата рождения пациента                      0
dtype: int64

In [363]:
# Подсчитаем кол-во уникальных значений внутри каждого признака.
for col in project_df.columns:
    print(f'Признак: {project_df[col].value_counts(dropna=False)}')
    print('-' * 60)

Признак: ID истории болезни
ffa72572-f065-11ec-ab56-0050568844e6    547
70076ff2-f7da-11ee-ab6f-0050568844e6    495
c1f452ff-d091-11ee-8606-005056880ecb    476
d13aae4f-a605-11ed-ab61-0050568844e6    466
aaf39b25-e684-11ee-ab6f-0050568844e6    429
                                       ... 
b86cafe6-6a6c-11ea-80cd-901b0e633689     26
4dd6bc50-2fe4-11ec-bb98-2cea7fe73f75     25
d13f240e-4d5b-11ea-80c8-901b0e63368a     24
5bcdbef9-efcf-11ed-8604-005056880ecb     23
70f64d47-d5c1-11e9-80bb-901b0e633689      1
Name: count, Length: 243, dtype: int64
------------------------------------------------------------
Признак: Осн. диаг. при выписке МКБ10 (текст)
Другой уточненный лейкоз                       22046
Острый миелоидный лейкоз                        5636
Другой миелоидный лейкоз                        1677
Хронический лимфоцитарный лейкоз                 925
Плазмоклеточный лейкоз                           543
Острый лимфобластный лейкоз                      404
Хронический миелоидный л

In [364]:
# Выведем названия всех признаков в таблице.
project_df.columns

Index(['ID истории болезни', 'Осн. диаг. при выписке МКБ10 (текст)',
       'Заголовок документа', 'Кол. лаб. показатель',
       'Значение кол. показателя', 'Ед. изм. кол. показателя',
       'Норма кол. показателя', 'Флаг нормы кол. показателя',
       'Кач. лаб. показатель', 'Значение кач. показателя',
       'Норма кач. показателя', 'Пол', 'Дата рождения пациента'],
      dtype='object')

**<u>Предварительные выводы по признакам:</u>**

1) `'ID истории болезни'`

    * Значения не полностью уникальны. Присутствуют разные анализы одного и того же пациента.

2) `'Осн. диаг. при выписке МКБ10'`

    * 11 уникальных значений. При этом среди них нет здоровых пациентов.

3) `'Заголовок документа'`

    * Всего одно уникальное значение. Неинформативный признак. Можно будет от него избавиться.

4) `'Кол. лаб. показатель'`

    * Очень большое кол-во уникальных значений. Связано скорее всего с тем, что один и тот же показатель записан по-разному. Потребуется перезаписать значения.

5) `'Значение кол. показателя'`

    * На данном этапе невозможно анализировать эти значения. Возможно, если у одного и того же показателя разные единицы измереня, потребуется перезаписать результаты.

6) `'Ед. изм. кол. показателя'`

    * Уже на этом этапе видно, что одни и те же единицы измереня записаны по-разному. Потребуется предобработка.
    * Есть пропуски.
    
7) `'Норма кол. показателя'`

    * В ходе обсуждения было выяснено, что даже для одного и того же показателя нормы различались. Возможно потребуется предобработка данного признака (заменить его на новые признаки `'нижняя граница нормы'` и `'верхняя граница нормы'`).

8) `'Флаг нормы кол. показателя'`

    * Признак, в котором записан результат сравнения показателя с референcной нормой.
    * Есть пропуски.

9) `'Кач. лаб. показатель'`

    * Преимущественное большинство это пропуски.
    * Непонятно, что обозначает значение *"Комментарий"*.
    * Возможно для нас это будет неинформативный признак.

10) `'Значение кач. показателя'`

    * Много пропусков.
    * Видимо в этом признаке раскрывается тот самый *"Комментарий"*.
    * Возможно для нас это будет неифнормативный признак.

11) `'Норма кач. показателя'`

    * Кол-во пропусков совпадает с кол-вом пропусков в предыдущих двух признаках.
    * Предварительные выводы такие же.

12) `'Пол'`

    * Всего два значения, имеет смысл перекодировать (Ж = 0, М = 1).

13) `'Дата рождения пациента'`

    * В том формате, в котором представлена, тяжело воспринимается. Имеет смысл извлечь возраст пациента и записать в новый признак, а от этого избавиться.

## Признак 'ID истории болезни'

In [365]:
# Присвоим каждому ID и/б своё число.
id_list = list(project_df['ID истории болезни'].unique())
project_df['ID истории болезни'] = project_df['ID истории болезни'].map(lambda x: id_list.index(x))

## Признак 'Заголовок документа'

In [366]:
# Избавимся от признака `Заголовок документа`.
project_df = project_df.drop(labels='Заголовок документа', axis=1)

## Признак 'Кол. лаб. показатель'

In [367]:
# Посмотрим все уникальные значения в признаке `Кол. лаб. показатель`.
project_df['Кол. лаб. показатель'].unique()

array(['Лейкоциты (WBC)', 'Эритроциты (RBC)', 'Гемоглобин (HGB)',
       'Гематокрит (HCT)', 'Средний объем эритроцита (MCV)',
       'Среднее содержание гемоглобина в эритроците (MCH)',
       'Средняя концентрация гемоглобина в эритроците (MCHC)',
       'Тромбоциты (PLT)', 'Нейтрофилы, относительное количество (NE%)',
       'Лимфоциты, относительное количество (LY%)',
       'Нейтрофилы, абсолютное количество (NE#)',
       'Лимфоциты, абсолютное количество (LY#)',
       'Смешанная фракция, абсолютное количество (MXD#)',
       'Смешанная фракция, относительное количество (MXD%)',
       'Ширина распределения эритроцитов по объему, коэффициент вариации (RDW-CV)',
       'Ширина распределения эритроцитов по объему, стандартное отклонение (RDW-SD)',
       'Ширина распределения эритроцитов по объему (RDW)',
       'Тромбокрит (PCT)', 'Средний объем тромбоцита (MPV)',
       'Ширина распределения тромбоцитов (PDW)',
       'Моноциты, относительное количество (MO%)',
       'Эозинофил

In [ ]:
# Перезапишем значения, чтобы сократить кол-во уникальных значений.
# Создадим функцию, которая будет проверять значение и перезаписывать его.
def new_name(value: object) -> str:
    """Функция для преобразования похожих значений в столбце 'Кол. лаб. показатель'
    
    Args: 
        значения типа object

    Returns:
       значения типа str
    """
    
    if value in ['Общее количество лейкоцитов (WBC)', 
                 'Лейкоциты (WBC)', 
                 'Лейкоциты']:
        return 'WBC'
    elif value in ['Общее количество эритроцитов (RBC)', 
                   'Эритроциты (RBC)', 
                   'Эритроциты']:
        return 'RBC'
    elif value in ['Гемоглобин (HGB)', 
                   'Гемоглобин']:
        return 'HGB'
    elif value in ['Гематокрит (HCT)', 
                   'Гематокрит']:
        return 'HCT'
    elif value in ['Тромбоциты (PLT)', 
                   'Тромбоциты']:
        return 'PLT'
    elif value in ['Тромбокрит (PCT)', 
                   'Тромбокрит']:
        return 'PCT'
    elif value in ['Ширина распределения эритроцитов (RDW)', 
                   'Ширина распределения эритроцитов', 
                   'Ширина распределения эритроцитов по объему (RDW)']:
        return 'RWD'
    elif value in ['Средний объем эритроцита (MCV)', 
                   'Средний объем эритроцита']:
        return 'MCV'
    elif value in ['Среднее содержание гемоглобина в эритроците (MCH)', 
                   'Среднее содержание гемоглобина в эритроците']:
        return 'MCH'
    elif value in ['Средняя концентрация гемоглобина в эритроците (MCHC)', 
                   'Средняя концентрация гемоглобина в эритроците']:
        return 'MCHC'
    elif value in ['Ширина распределения эритроцитов по объему, стандартное отклонение (RDW-SD)',
                   'Ширина распределения эритроцитов, стандартное отклонение (RDW-SD)',
                   'Ширина распределения эритроцитов по объему, стандартное отклонение']:
        return 'RDW-SD'
    elif value in ['Ширина распределения эритроцитов по объему, коэффициент вариации (RDW-CV)',
                   'Ширина распределения эритроцитов по объему, коэффициент вариации']:
        return 'RDW-CV'
    elif value in ['MPV (Средний объём тромбоцитов)', 
                   'Средний объем тромбоцита (MPV)', 
                   'Средний объём тромбоцитов', 
                   'Средний объем тромбоцита']:
        return 'MPV'
    elif value in ['Ширина распределения тромбоцитов (PDW)', 
                   'Ширина распределения тромбоцитов по объему', 
                   'Ширина распределения тромбоцитов']:
        return 'PDW'
    elif value in ['Ретикулоциты %']:
        return 'RET%'
    elif value in ['Ретикулоциты кол-во']:
        return 'RET#'
    elif value in ['СОЭ Панч', 
                   'Скорость оседания эритроцитов (СОЭ)', 
                   'СОЭ Панч.', 
                   'СОЭ Вест.', 
                   'Скорость оседания эритроцитов (СОЭ) по Вестергрену', 
                   'СОЭ по Панченкову']:
        return 'ESR'
    elif value in ['Цветовой показатель']:
        return 'CI'
    elif value in ['Относительное количество нейтрофилов (NE%)', 
                   'Нейтрофилы, относительное количество (NE%)',
                   'Нейтрофилы, относительное количество', 
                   'Нейтрофилы %']:
        return 'NE%'
    elif value in ['Сегментоядерные']:
        return 'NE%S'
    elif value in ['Юные']:
        return 'NE%YY'
    elif value in ['Палочкоядерные']:
        return 'NE%Y'
    elif value in ['Относительное количество лимфоцитов (LY%)', 
                   'Лимфоциты, относительное количество (LY%)', 
                   'Лимфоциты, относительное количество', 'Лимфоциты %']:
        return 'LY%'
    elif value in ['Относительное количество базофилов (BA%)', 
                   'Базофилы, относительное количество (BA%)', 
                   'Базофилы, относительное количество', 
                   'Базофилы %', 
                   'Базофилы']:
        return 'BA%'
    elif value in ['Относительное количество моноцитов (MO%)', 
                   'Моноциты, относительное количество (MO%)', 
                   'Моноциты, относительное количество', 
                   'Моноциты %']:
        return 'MO%'
    elif value in ['Моноциты']:
        return 'MO'
    elif value in ['Относительное количество эозинофилов (EO%)',
                   'Эозинофилы, относительное количество (EO%)', 
                   'Эозинофилы, относительное количество', 
                   'Эозинофилы %']:
        return 'EO%'
    elif value in ['Эозинофилы']:
        return 'EO'
    elif value in ['Смешанная фракция, относительное количество (MXD%)', 
                   'Смешанная фракция, относительное количество', 
                   'Неклассифицируемые %', 
                   'MXD']:
        return 'MDX%'
    elif value in ['Абсолютное количество нейтрофилов (NE#)', 
                   'Нейтрофилы, абсолютное количество (NE#)', 
                   'Нейтрофилы, абсолютное количество', 
                   'Нейтрофилы #']:
        return 'NE#'
    elif value in ['Абсолютное количество лимфоцитов (LY#)', 
                   'Лимфоциты, абсолютное количество (LY#)', 
                   'Лимфоциты #', 
                   'Лимфоциты, абсолютное количество']:
        return 'LY#'
    elif value in ['Лимфоциты']:
        return 'LY'
    elif value in ['Абсолютное количество базофилов (BA#)', 
                   'Базофилы, абсолютное количество (BA#)', 
                   'Базофилы, абсолютное количество', 
                   'Базофилы #', 
                   'Базофилы # ']:
        return 'BA#'
    elif value in ['Абсолютное количество моноцитов (MO#)', 
                   'Моноциты, абсолютное количество (MO#)', 
                   'Моноциты #', 
                   'Моноциты, абсолютное количество']:
        return 'MO#'
    elif value in ['Абсолютное количество эозинофилов (EO#)', 
                   'Эозинофилы, абсолютное количество (EO#)', 
                   'Эозинофилы, абсолютное количество', 
                   'Эозинофилы # ']:
        return 'EO#'
    elif value in ['Смешанная фракция, абсолютное количество (MXD#)', 
                   'Смешанная фракция, абсолютное количество', 
                   'Неклассифицируемые кол-во', 
                   'MXD# ']:
        return 'MDX#'
    elif value in ['Плазматич. клетки', 
                   'Плазматические клетки', 
                   'Миелоциты', 
                   'Промиелоциты', 
                   'Пролимфоциты', 
                   'Бласты',
                   'Нормобласты', 
                   'Нормобласты #', 
                   'Нормобласты %']:
        return 'Atypical cells'
    else:
        return value

In [369]:
# Применим эту функцию.
project_df['Кол. лаб. показатель'] = project_df['Кол. лаб. показатель'].apply(new_name)

In [370]:
# Посмотрим, что получилось после обработки.
project_df['Кол. лаб. показатель'].value_counts(dropna=False)

Кол. лаб. показатель
RBC               1594
PLT               1553
HGB               1517
WBC               1512
MCHC              1489
HCT               1451
LY%               1447
NE%               1310
MCV               1267
LY#               1264
NE#               1247
MCH               1075
BA%               1061
MO%                803
PCT                760
EO%                748
ESR                738
MO#                721
LY                 719
CI                 716
NE%S               715
MPV                695
MO                 666
Atypical cells     663
MDX%               646
BA#                640
EO#                619
PDW                600
NE%Y               575
EO                 533
RDW-SD             513
RWD                413
MDX#               395
RDW-CV             326
RDW                295
NE%YY              276
MPV                139
MXD#                96
RDW-CV              90
MXD%                89
P-LCR               78
RET%                57
NaN          

In [371]:
# # Посмотрим на 51 пациента с пропусками в данном признаке.
# project_df[project_df[['Кол. лаб. показатель']].isna().all(axis=1)]

У всех пациентов с пропуском в признаке `'Кол. лаб. показатель'` отсутствуют значения в других важных признаках, на основе которых можно было бы предложить, что это за показатель. Заполнить эти пропуски мы не сможем, поэтому имеет смысл избавиться от данных строчек.

In [372]:
# Удаляем строки с пропусками в 'Кол. лаб. показатель'.
project_df = project_df.dropna(subset = ['Кол. лаб. показатель'])
project_df.shape

(32155, 12)

## Признак 'Ед. изм. кол. показателя'

In [ ]:
# Напишем функцию, которая заменяет различные варианты написания 'фл'.
def new_units(value: object) -> str:
    """Функция для преобразования похожих значений в столбце 'Ед. изм. кол. показателя'
    
    Args: 
        значения типа object

    Returns:
       значения типа str
    """
    
    if value in ['fL', 'FL']:
        return 'фл'
    else:
        return value

In [374]:
# Применим эту функцию.
project_df['Ед. изм. кол. показателя'] = project_df['Ед. изм. кол. показателя'].apply(new_units)

In [375]:
# Посмотрим, что получилось после обработки.
project_df['Ед. изм. кол. показателя'].value_counts(dropna=False)

Ед. изм. кол. показателя
%          13627
10*9/л      8070
фл          3122
г/л         2384
10*12/л     1594
пг          1075
_            918
мм/час       738
г/дл         627
Name: count, dtype: int64

In [376]:
# Проверим одинаковые ли системы измерения у одного и того же показателя.
units_of_measurement = project_df.groupby('Кол. лаб. показатель')['Ед. изм. кол. показателя'].nunique().sort_values()
units_of_measurement[units_of_measurement > 1]

Кол. лаб. показатель
MPV               2
MXD#              2
MCHC              2
PDW               3
Atypical cells    3
Name: Ед. изм. кол. показателя, dtype: int64

In [377]:
# Выведем различия в измерениях для каждого значения.
check_units = project_df[project_df['Кол. лаб. показатель'].isin(units_of_measurement[units_of_measurement > 1].index)]
check_units.groupby('Кол. лаб. показатель')['Ед. изм. кол. показателя'].value_counts()

Кол. лаб. показатель  Ед. изм. кол. показателя
Atypical cells        %                           487
                      _                           164
                      10*9/л                       12
MCHC                  г/л                         862
                      г/дл                        627
MPV                   фл                          690
                      _                             5
MXD#                  10*9/л                       63
                      _                            33
PDW                   фл                          513
                      %                            82
                      г/л                           5
Name: count, dtype: int64

1) `'Atypical cells'`: совершенно разные системы измерения, нужно будет определиться с какой будем дальше работать, т.к. вручную мы это никак не преобразуем.

2) `'MCHC'`: децилитры перевести в литры мы сможем.

3) `'MPV'`: всегда измеряется в фемтолитрах (фл). Нужно проверить 5 показателей со значением измерения `'_'`.

4) `'MXD#'`: нужно проверить 33 показателя со значением измереня `'_'`, скорее всего это та же система измереня.

5) `'PDW'`: совершенно разные системы измерения, нужно будет определиться с какой будем дальше работать, т.к. вручную мы это никак не преобразуем.

In [378]:
# У 12 пациентов с 'Atypical cells' и 'Ед. изм. кол. показателя' = '10*9/л', результат = 0.
# Поэтому их можно преобразовать в любую другую систему измерения.
# Изменим данные этих пациентов под преобладающую систему измерения.
project_df.loc[(project_df['Кол. лаб. показатель'] == 'Atypical cells')&(project_df['Ед. изм. кол. показателя'] == '10*9/л'), 'Норма кол. показателя'] = '0:0'
project_df.loc[(project_df['Кол. лаб. показатель'] == 'Atypical cells')&(project_df['Ед. изм. кол. показателя'] == '10*9/л'), 'Ед. изм. кол. показателя'] = '%'

In [379]:
# Преобразованиями для 'MCHC'.
project_df.loc[(project_df['Кол. лаб. показатель'] == 'MCHC')&(project_df['Ед. изм. кол. показателя'] == 'г/дл'), 'Норма кол. показателя'] = '300:380'
project_df.loc[(project_df['Кол. лаб. показатель'] == 'MCHC')&(project_df['Ед. изм. кол. показателя'] == 'г/дл'), 'Значение кол. показателя'] = project_df.loc[(project_df['Кол. лаб. показатель'] == 'MCHC')&(project_df['Ед. изм. кол. показателя'] == 'г/дл'), 'Значение кол. показателя'] * 0.1
project_df.loc[(project_df['Кол. лаб. показатель'] == 'MCHC')&(project_df['Ед. изм. кол. показателя'] == 'г/дл'), 'Ед. изм. кол. показателя'] = 'г/л'

In [380]:
# Преобразования для 'MPV'.
project_df.loc[(project_df['Кол. лаб. показатель'] == 'MPV')&(project_df['Ед. изм. кол. показателя'] == '_'), 'Норма кол. показателя'] = '6,5:12,6'
project_df.loc[(project_df['Кол. лаб. показатель'] == 'MPV')&(project_df['Ед. изм. кол. показателя'] == '_'), 'Ед. изм. кол. показателя'] = 'фл'

In [381]:
# Преобразования для 'MXD#'.
project_df.loc[(project_df['Кол. лаб. показатель'] == 'MXD#')&(project_df['Ед. изм. кол. показателя'] == '_')&(project_df['Значение кол. показателя'] > 0.8), 'Флаг нормы кол. показателя'] = 'Повыш'
project_df.loc[(project_df['Кол. лаб. показатель'] == 'MXD#')&(project_df['Ед. изм. кол. показателя'] == '_')&(project_df['Значение кол. показателя'] < 0.2), 'Флаг нормы кол. показателя'] = 'Пониж'
project_df.loc[(project_df['Кол. лаб. показатель'] == 'MXD#')&(project_df['Ед. изм. кол. показателя'] == '_'), 'Норма кол. показателя'] = '0,2:0,8'
project_df.loc[(project_df['Кол. лаб. показатель'] == 'MXD#')&(project_df['Ед. изм. кол. показателя'] == '_'), 'Ед. изм. кол. показателя'] = '10*9/л'

## Признак 'Норма кол. показателя'

In [382]:
# Проверяем нормы для различных показателей.
norms_of_values = project_df.groupby('Кол. лаб. показатель')['Норма кол. показателя'].nunique()
check_norms = project_df[project_df['Кол. лаб. показатель'].isin(norms_of_values[norms_of_values > 1].index)]
all_norms = check_norms.groupby(['Кол. лаб. показатель'])['Норма кол. показателя'].value_counts()
all_norms

Кол. лаб. показатель  Норма кол. показателя
Atypical cells        0:0                       477
                      0:2                       126
                      0:1                        48
                      _                          12
BA#                   0:0,1                     622
                                               ... 
RWD                   10,5:18,0                   3
WBC                   4:11                     1429
                      4.0:8.8                    41
                      4,00:11,00                 23
                      4:8,8                      19
Name: count, Length: 131, dtype: int64

In [383]:
all_norms.index.values

array([('Atypical cells', '0:0'), ('Atypical cells', '0:2'),
       ('Atypical cells', '0:1'), ('Atypical cells', '_'),
       ('BA#', '0:0,1'), ('BA#', '0,00:0,10'), ('BA#', '0:0.1'),
       ('BA%', '0:2'), ('BA%', '0,0:2,0'), ('CI', '0,8:1,05'),
       ('CI', '0,85:1,15'), ('CI', '_'), ('CI', '0,8:1,5'),
       ('CI', '0.8:1.05'), ('CI', '0,80:1,05'), ('CI', '0.80:1.05'),
       ('EO#', '0:0,4'), ('EO#', '0,00:0,40'), ('EO%', '0:5'),
       ('EO%', '0,0:5,0'), ('ESR', '2:20'), ('ESR', 'Общая норма'),
       ('ESR', '1:20'), ('ESR', '2:15'), ('HCT', '35:47'),
       ('HCT', '39:52'), ('HCT', '35:52'), ('HCT', '34:46'),
       ('HCT', '30.0:49.5'), ('HCT', '35,0:47,0'), ('HCT', '39,0:52,0'),
       ('HGB', '117:160'), ('HGB', '132:180'), ('HGB', '117:180'),
       ('HGB', '117:140'), ('HGB', '114:154'), ('HGB', '132:160'),
       ('HGB', '122:168'), ('LY#', '1:3,7'), ('LY#', '_'),
       ('LY#', '1,00:3,70'), ('LY#', '1,00:3.70'), ('LY#', '1,0:3,7'),
       ('LY%', '18:44'), ('LY%', '2

## Признак 'Флаг нормы кол. показателя'

In [384]:
# Посмотрим значения.
project_df['Флаг нормы кол. показателя'].value_counts(dropna=False)

Флаг нормы кол. показателя
Норм     18497
Пониж     7214
Повыш     6381
Норма       63
Name: count, dtype: int64

In [385]:
# Уберем предупреждения, чтобы они не загромождали вывод.
import warnings
warnings.filterwarnings('ignore')

In [386]:
# Заменим значения.
project_df['Флаг нормы кол. показателя'].replace('Норм', 'Норма', inplace=True)

## Признаки 'Кач. лаб. показатель', 'Значение кач. показателя', 'Норма кач. показателя'

Т.к. преимущественное большинство значений в данных признаках пропуски, а оставшиеся значения носят индивидуальный характер (пояснения к конкретному анализу), данные признаки неинформативны.

In [387]:
# Избавимся от данных признаков.
project_df = project_df.drop(labels='Кач. лаб. показатель', axis=1)
project_df = project_df.drop(labels='Значение кач. показателя', axis=1)
project_df = project_df.drop(labels='Норма кач. показателя', axis=1)

## Признак 'Пол'

In [388]:
# Изменим значения в данном столбце.
project_df['Пол'] = project_df['Пол'].map({'Ж': 0, 'М': 1})

## Признак 'Дата рождения пациента'

In [389]:
# Импорт библиотеки для работы с датой и временем
from datetime import datetime, date

In [390]:
# Преобразование признака в формат даты.
project_df['Дата рождения пациента'] = pd.to_datetime(project_df['Дата рождения пациента'])

In [ ]:
# Функция для высчитывания возраста.
def age(date_of_born: object) -> int: 
    """Функция для извлечения возраста пациента из столбца 'Дата рождения пациента'
    
    Args: 
        значения типа object

    Returns:
       значения типа int
    """
    
    today = date.today() 
    return today.year - date_of_born.year - ((today.month, today.day) < (date_of_born.month, date_of_born.day))

In [392]:
# Применим функцию.
project_df['Возраст'] = project_df['Дата рождения пациента'].apply(age)

In [393]:
# Удалим столбец 'Дата рождения пациента'.
project_df = project_df.drop(labels='Дата рождения пациента', axis=1)

In [394]:
# Переименуем столбцы для удобства.
project_df.rename(columns={
    'ID истории болезни': 'ID пациента',
    'Осн. диаг. при выписке МКБ10 (текст)': 'Диагноз',
    'Кол. лаб. показатель': 'Лабораторный показатель',
    'Значение кол. показателя': 'Значение',
    'Ед. изм. кол. показателя': 'Единицы измерения',
    'Норма кол. показателя': 'Референсный интервал',
    'Флаг нормы кол. показателя': 'Результат'
    },
    inplace=True
)

In [395]:
# Результат преобразований.
project_df

ID пациента                                      Диагноз  \
0                0                       Плазмоклеточный лейкоз   
1                0                       Плазмоклеточный лейкоз   
2                0                       Плазмоклеточный лейкоз   
3                0                       Плазмоклеточный лейкоз   
4                0                       Плазмоклеточный лейкоз   
...            ...                                          ...   
36292          242  Острый лейкоз неуточненного клеточного типа   
36293          242  Острый лейкоз неуточненного клеточного типа   
36294          242  Острый лейкоз неуточненного клеточного типа   
36295          242  Острый лейкоз неуточненного клеточного типа   
36296          242  Острый лейкоз неуточненного клеточного типа   

      Лабораторный показатель  Значение Единицы измерения  \
0                         WBC      1.80            10*9/л   
1                         RBC      4.18           10*12/л   
2                         HGB    120.00               г/л   
3                         HCT     35.10                 %   
4                         MCV     84.00                фл   
...                       ...       ...               ...   
36292                     BA%      1.00                 %   
36293                      EO      0.00                 %   
36294          Atypical cells      0.00                 %   
36295                      LY     65.00                 %   
36296                      MO      1.00                 %   

      Референсный интервал Результат  Пол  Возраст  
0                     4:11     Пониж    0       62  
1                  3,8:5,7     Норма    0       62  
2                  117:160     Норма    0       62  
3                    35:47     Норма    0       62  
4                    80:99     Норма    0       62  
...                    ...       ...  ...      ...  
36292                  0:2     Норма    1       10  
36293                  0:5     Норма    1       10  
36294                  0:0     Норма    1       10  
36295                18:44     Повыш    1       10  
36296                 2:12     Пониж    1       10  

[32155 rows x 9 columns]

In [396]:
# Посмотрим информацию о признаках.
project_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 32155 entries, 0 to 36296
Data columns (total 9 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   ID пациента              32155 non-null  int64  
 1   Диагноз                  32155 non-null  object 
 2   Лабораторный показатель  32155 non-null  object 
 3   Значение                 32155 non-null  float64
 4   Единицы измерения        32155 non-null  object 
 5   Референсный интервал     32155 non-null  object 
 6   Результат                32155 non-null  object 
 7   Пол                      32155 non-null  int64  
 8   Возраст                  32155 non-null  int64  
dtypes: float64(1), int64(3), object(5)
memory usage: 2.5+ MB


In [397]:
# Выведем кол-во пропущенных значений в каждом признаке.
project_df.isna().sum()

ID пациента                0
Диагноз                    0
Лабораторный показатель    0
Значение                   0
Единицы измерения          0
Референсный интервал       0
Результат                  0
Пол                        0
Возраст                    0
dtype: int64

Было решено выбрать таргетом признак `'Результат'`. В связи с чем:

1) Признак `'ID пациента'` не несёт никакой информации и будет лучше его удалить.

2) Признак `'Диагноз'` будет вносить шум. Т.к. 100% пациентов в таблице с тем или иным диагнозом лейкоза (здоровых нет), несмотря на результаты анализа, т.е. этот диагноз закреплён на ID пациента и остаётся за ним, даже если по анализам у пациента ремиссия, для нас он неинформативен. Также будет удалён.

3) Признак `'Референсный интервал'` лучше разделить на 2 новых признака: `'Верхняя граница`' и `'Нижняя граница'`, а от самого признака избавиться.

4) Обновить информацию в признаке `'Результат'`, т.к. в таблице встречались ошибки, путём сравнения `'Значение'` c 2 новыми признаками (`'Верхняя граница`' и `'Нижняя граница'`) и получить 2 значения в данном признаке: 'Норма' = 0, 'Отклонение' = 1.

In [398]:
# 1) Удаляем 2 неинформативных признака.
project_df = project_df.drop(labels='ID пациента', axis=1)
project_df = project_df.drop(labels='Диагноз', axis=1)

In [399]:
# Проверим уникальные значения в признаке 'Референсный интервал'.
project_df['Референсный интервал'].unique()

array(['4:11', '3,8:5,7', '117:160', '35:47', '80:99', '27:38', '300:380',
       '150:450', '45:72', '18:44', '1,5:7', '1:3,7', '0,2:0,8', '5:10',
       '10,5:18', '37:54', '0,19:0,36', '9,1:12,6', '9,3:16,7', '2:12',
       '0:5', '0:2', '0,1:0,7', '0:0,4', '0:0,1', '0,8:1,05', '2:20',
       '1:6', '0:0', '6,5:10', '9:17', '4,3:6,1', '132:180', '39:52',
       '10:18', '6,5:12,6', '0,2:2', '17:70', '0,85:1,15', '0:4', '2:10',
       '4,30:6,10', '4,00:11,00', '45,0:72,0', '18,0:44,0 ', '1,50:7,00',
       '1,00:3,70', '0.2:0.8', '_', '0,10:0,70', '0,00:0,40', '0,00:0,10',
       '18,0:44,0', '2,0:12,0', '0,0:5,0', '0,0:2,0', '3,8:6,1',
       '117:180', '35:52', '0:0,7', '0:1', '1:20', '2,0:12,0 ',
       '1,00:3.70', '0,2:2,0', '117:140', '4.0:8.8', '3.5:5.2', '114:154',
       '34:46', '26.5:35.5', '32.2:35.6', '150:320', '41:75', '20:45',
       '7.5:10.0', '15.0:17.0', '3:8', '10.7:45', '4:8,8', '3,5:5,2',
       '26,5:35,5', '32,2:35,6', '7,5:10', '15:17', '10,7:45',
       '0

In [400]:
project_df[project_df['Референсный интервал'] == '_'].shape

(121, 7)

In [401]:
project_df[project_df['Референсный интервал'] == 'Общая норма'].shape

(105, 7)

Просмотрев все нормы для разных лабораторных показателей (строка: [121]; код: `all_norms.index.values`), можно сделать вывод, что референсные интервалы одинаковые для всех пациентов только для 2 лабораторных показателей (`'LY#'`, `'NE#'`). В остальных случаях есть несколько интервалов. Как известно, они зависят не только от пола и возраста, но и от аппарата, на котором проводились измерения. А так как мы работаем с медицинскими данными, то ошибок быть не должно. Применять моду для замены значений будет неправильно, следовательно, их нужно удалить.

In [402]:
# Заменяем значения, где можем.
project_df.loc[(project_df['Лабораторный показатель'] == 'LY#')&(project_df['Референсный интервал'] == '_'), 'Референсный интервал'] = '1:3.7'
project_df.loc[(project_df['Лабораторный показатель'] == 'NE#')&(project_df['Референсный интервал'] == '_'), 'Референсный интервал'] = '1.5:7'

In [403]:
# Удаляем неправильные референсыные интервалы.
project_df.drop(project_df.index[project_df['Референсный интервал'] == '_'], inplace=True)
project_df.drop(project_df.index[project_df['Референсный интервал'] == 'Общая норма'], inplace=True)

In [404]:
# Посмотрим, сколько наблюдений было удалено.
project_df.shape

(32007, 7)

In [405]:
# 2) Разделяем 'Референсный интервал'.
project_df[['Нижняя граница', 'Верхняя граница']] = project_df['Референсный интервал'].str.split(':', expand=True)

project_df['Нижняя граница'] = project_df['Нижняя граница'].str.replace(',', '.')
project_df['Верхняя граница'] = project_df['Верхняя граница'].str.replace(',', '.')

project_df['Нижняя граница'] = project_df['Нижняя граница'].astype(float)
project_df['Верхняя граница'] = project_df['Верхняя граница'].astype(float)

project_df = project_df.drop(labels='Референсный интервал', axis=1)

In [406]:
# Изменим порядок столбцов, для удобства восприятия.
project_df = project_df[['Пол', 'Возраст', 'Лабораторный показатель', 'Значение', 'Единицы измерения', 'Нижняя граница', 'Верхняя граница', 'Результат']]
# Изменим индексы на новые.
project_df.reset_index(drop=True, inplace=True)

In [407]:
project_df

Пол  Возраст Лабораторный показатель  Значение Единицы измерения  \
0        0       62                     WBC      1.80            10*9/л   
1        0       62                     RBC      4.18           10*12/л   
2        0       62                     HGB    120.00               г/л   
3        0       62                     HCT     35.10                 %   
4        0       62                     MCV     84.00                фл   
...    ...      ...                     ...       ...               ...   
32002    1       10                     BA%      1.00                 %   
32003    1       10                      EO      0.00                 %   
32004    1       10          Atypical cells      0.00                 %   
32005    1       10                      LY     65.00                 %   
32006    1       10                      MO      1.00                 %   

       Нижняя граница  Верхняя граница Результат  
0                 4.0             11.0     Пониж  
1                 3.8              5.7     Норма  
2               117.0            160.0     Норма  
3                35.0             47.0     Норма  
4                80.0             99.0     Норма  
...               ...              ...       ...  
32002             0.0              2.0     Норма  
32003             0.0              5.0     Норма  
32004             0.0              0.0     Норма  
32005            18.0             44.0     Повыш  
32006             2.0             12.0     Пониж  

[32007 rows x 8 columns]